core/src/main/java/org/elasticsearch/common/util/CountedBitSet.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.FixedBitSet;
+
+/**
+ * A {@link CountedBitSet} wraps a {@link FixedBitSet} but automatically releases the internal bitset
+ * when all bits are set to reduce memory usage. This structure can work well for sequence numbers
+ * from translog as these numbers are likely to form contiguous ranges (eg. filling all bits).
+ */
+public final class CountedBitSet extends BitSet {
```

Comment:
This only has one consumer, does it really need to be in a common package? What's the motivation for moving this out of the translog package?

Reply:
Boaz suggested to make it as a drop-in replacement for `BitSet`. I thought about the `LocalCheckpointTracker` and moved it to the common package. I am happy to move it back to the translog package.

Reply:
First, the description of the PR is "Move CountedBitSet to the common package" yet it appears the changes are larger than that; it's best keep the summary and the changes aligned. Also, I think moving and changing the implementation should be separate PRs and I think the reason for the move should be explained in the commit message.

With that out of the way: if the intention is to reuse this in the local checkpoint tracker I still do not think these needs to be in common, it can be in the sequence number package `org.elasticseach.index.seqno` since the only uses relate to sequence numbers.  It's a shame this class has to be public in that package even.

My resistance is that the common package is already a bit of a dumping ground and at least in the sequence number package it's slightly clearer than it's an internal implementation rather than for public consumption (e.g., in plugins). When we modularize (JDK modules) we will be in a better position to make clear what is part of the public API and what is not.

Reply:
> First, the description of the PR is "Move CountedBitSet to the common package" yet it appears the changes are larger than that; it's best keep the summary and the changes aligned.

My bad, I should have updated the description more carefully when addressing feedbacks.

>  Also, I think moving and changing the implementation should be separate PRs

I agree. I will move this class back to the `translog` package in this PR. If we agree to use it in the `LocalCheckpointTracker`, we can make another PR to move it the `seqno` package later.

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/27547/commits/0df047404b95ba99ce6fbcc0cf9c760053de0de8